mlir/test/Dialect/Arith/int-range-interface.mlir

```diff
@@ -266,6 +266,18 @@ func.func @remui_base(%arg0 : index, %arg1 : index ) -> i1 {
     func.return %3 : i1
 }
 
+// CHECK-LABEL: func @remui_base_maybe_zero
+// CHECK: %[[true:.*]] = arith.constant true
+// CHECK: return %[[true]]
+func.func @remui_base_maybe_zero(%arg0 : index, %arg1 : index ) -> i1 {
+    %c4 = arith.constant 4 : index
+
+    %0 = arith.minui %arg1, %c4 : index
+    %1 = arith.remui %arg0, %0 : index
+    %2 = arith.cmpi ult, %1, %c4 : index
+    func.return %2 : i1
+}    
```

Comment:
Having checked, [alive2 is happy with this, even in its general form](https://alive2.llvm.org/ce/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:llvm,selection:(endColumn:16,endLineNumber:21,positionColumn:16,positionLineNumber:21,selectionStartColumn:16,selectionStartLineNumber:21,startColumn:16,startLineNumber:21),source:'declare+i8+@llvm.umin.i8(i8,+i8)%0A%0Adefine+i1+@src_drop_urem(i8+%25x,+i8+%25y)+%7B%0A++%25clamp+%3D+call+i8+@llvm.umin.i8(i8+%25x,+i8+4)%0A++%25mod+%3D+urem+i8+%25y,+%25clamp%0A++%25ret+%3D+icmp+ult+i8+%25mod,+4%0A++ret+i1+%25ret%0A%7D%0A%0Adefine+i1+@tgt_drop_urem(i8+%25x,+i8+%25y)+%7B%0A++ret+i1+1%0A%7D%0A%0Adefine+i1+@src_drop_urem_noconst(i8+%25x,+i8+%25y,+i8+%25z)+%7B%0A++%25clamp+%3D+call+i8+@llvm.umin.i8(i8+%25x,+i8+%25z)%0A++%25mod+%3D+urem+i8+%25y,+%25clamp%0A++%25ret+%3D+icmp+ult+i8+%25mod,+%25z%0A++ret+i1+%25ret%0A%7D%0A%0Adefine+i1+@tgt_drop_urem_noconst(i8+%25x,+i8+%25y,+i8+%25z)+%7B%0A++ret+i1+1%0A%7D'),l:'5',n:'0',o:'LLVM+IR+source+%231',t:'0')),k:30.812101910828027,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:alive,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:llvm,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'alive-tv+(Editor+%231,+Compiler+%231)+LLVM+IR',t:'0')),k:69.18789808917197,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
What do you mean by that? And sorry for no alive2 link. Didn't know it was used by MLIR (given that alive2 verifies llvm-ir).

Reply:
I mean that it seems like this specific test case doesn't need int range analysis to prove the fold it safe -- we could do it as a canonicalization by matching the operation dag only.

I'm asking/suggesting to make it a more general canonicalization pattern outside of the int range analysis framework.

Reply:
Or this class of folds that we could do without reasoning about int ranges and only with simple dag matching.

Reply:
Would something like changing the cmp from `%c4` to a new constant work to move this outside of DAG range?

Reply:
I don't understand. The alive log shows that the second argument to cmp can be arbitrary.

Reply:
This is a niche and expensive enough pattern match that I wouldn't make it a canonicalization - especially since we're using `minui` to stand in for any number of things that would impose a maximum on a value.

The following example, which wouldn't make sense as a canonicalization, should also fold
```mlir
func.func @f(%arg0 : index) -> i1 {
  %tid = gpu.thread_id x upper_bound 64 : index
  %clamped = arith.remui %arg0, %tid : index
  %c64 = arith.constant 64 : index
  %ret = arith.cmpi ult %clamped, %c64
  func.return %ret : i1
}
```
which should simplify to true

Reply:
> The following example, which wouldn't make sense as a canonicalization

Why wouldn't you want to fold this as canonicalization? I didn't follow here...

Reply:
Because you don't know that the thread ID is <= 64 without running integer range analysis

Reply:
NB: Likewise for the current fold, but I'm happy to switch the above example if reviewers prefer.

Reply:
Re "why not a canonicalization" - in LLVM terms, this feels like an `InstCombine` type of rewrite and not an `InstSimplify` one, and `-canonicalize` is meant to be rather cheap like `InstSimplify`.